### ISSUE TITLE
Simplified kubectl access to private cluster

### ISSUE THREAD

**Issue by palma21** [Azure/AKS#2059]

**Comment by lgmorand** [comment]
That looks awesome. Any idea of the implementation ? How would it work, like what would be the user experience of that ?

**Comment by palma21** [comment]
https://docs.microsoft.com/en-us/azure/aks/private-clusters#aks-run-command-preview

**Comment by ohorvath** [comment]
That's a pretty useful feature, thank you. Are you going to support service principal based AKS clusters in the GA version?

**Comment by jabbera** [comment]
We need SP based support as well,

**Comment by palma21** [comment]
Should work on any kind of clusters SPN or managed identity
(but I do recommend you migrate from SP to managed identity, should be quick and painless otherwise let us know!)

**Comment by ohorvath** [comment]
@palma21 I tried again and it's failing with both SP and with User Assigned identities. I have Azure AD integration and RBAC. Also the AKS cluster creation is broken with user assigned identities, I've opened a support ticket for that part.

**Comment by miwithro** [comment]
@haitch can you look into this?

**Comment by haitch** [comment]
@ohorvath this is a internal error we are hitting, it is already fixed, our deployment pipeline will take another week to reach to westeurope.

**Comment by ohorvath** [comment]
@haitch Thanks for the update. Which error are you referring to? We have multiple issues currently. :) And I have my resources in westus2, centralus, so if the fix are there, I can try again.

**Comment by ohorvath** [comment]
I've checked this again and I can repro the issue. AKS clusters with BYOD user assigned identities work fine with the run command. However AAD integration break this feature. 

Failed to connect to MSI. Please make sure MSI is configured correctly.
Get Token request returned: <Response [400]>

I've changed nothing but the AAD integration.

**Comment by haitch** [comment]
@ohorvath I am pretty sure you are using in azure cloud shell.  This is a known issue from cloud shell, I can find you the link to this issue.

**Comment by haitch** [comment]
here is the issue tracking from cloudshell side:

https://github.com/Azure/azure-cli/issues/11749 
https://github.com/Azure/azure-cli/issues/17695

**Comment by shreyasHpandya** [comment]
Is it possible to call k8 API via this feature?

**Comment by haitch** [comment]
@shreyasHpandya you can run kubectl and helm in this cli, kubectl and helm invokes k8s API behind the sense.

so it is possible.

but my question is why? this feature is designed for cluster admin to run simple, one off command without worrying isolated network (given you are authorized in ARM and have permission).

If you are going to have some app and work with k8s API, then why bother going through ARM, you can directly deploy this app to your k8s cluster.

**Comment by shreyasHpandya** [comment]
Let's say I am developing a third-party CSPM tool that wants to scan their cluster for misconfigurations etc. my client has already given me access to their azure API, I want to use the same SP to extend functionality to AKS

**Comment by haitch** [comment]
that's not a designed scenario.

you should publish your app somewhere, then ask your client to install your app.

**Comment by shreyasHpandya** [comment]
Yes we do that currently, but this way we can reduce the pain of onboarding for client

**Comment by haitch** [comment]
for a normal cluster, if your client give you enough access to their AKS resource, then you should be able to call kubernetes API.

for a private cluster:
  - If you client give you access to both AKS resource and vNet resource, then you should also be able to call Kubernetes API from that vNet.
  - If your client give you only access to AKS resource without vNet resource, then you are cannot reach to Kubernetes API (this is the meaning of private cluster) 
  - with run-command preview feature turned on, when your client give you access to AKS resource, it allow you to access kubernetes API with limited kubectl command.

**Comment by mikebranstein** [comment]
I'm in agreement with @haitch. We have private clusters for a reason - we want to restrict access to internal networks only. I'm not opposed to this feature, but I wouldn't want the feature to be enabled by default. 

> for a normal cluster, if your client give you enough access to their AKS resource, then you should be able to call kubernetes API.
> 
> for a private cluster:
> 
> * If you client give you access to both AKS resource and vNet resource, then you should also be able to call Kubernetes API from that vNet.
> * If your client give you only access to AKS resource without vNet resource, then you are cannot reach to Kubernetes API (this is the meaning of private cluster)
> * with run-command preview feature turned on, when your client give you access to AKS resource, it allow you to access kubernetes API with limited kubectl command.

**Comment by phillebaba** [comment]
Is there any documentation for how this feature can be disabled in AKS?

**Comment by haitch** [comment]
CLI will be ready shortly, but you can use API to disable it for now:

https://github.com/Azure/azure-rest-api-specs/blob/191a721de644cc3f872f4fe9d676cf366083a106/specification/containerservice/resource-manager/Microsoft.ContainerService/stable/2021-08-01/managedClusters.json#L3901

https://github.com/Azure/azure-rest-api-specs/blob/main/specification/containerservice/resource-manager/Microsoft.ContainerService/stable/2021-08-01/examples/ManagedClustersCreate_DisableRunCommand.json
